### DISCUSSION TITLE
MathJax `[mml]/mml3` implementation

### DISCUSSION THREAD

**Discussion by ykla** [vuejs/vitepress#3917] [Q&A]
I merged someone else's output patch about markdown-it-mathjax3 (to make it output as text, so that doing so others can copy formulas) and deployed it. The table and text are displayed, but it seems to be missing the fonts MJXZERO, MJXTEX. but I can't find where these 2 fonts are, not sure if it's caused by other issues. Please help me to confirm. Thank you very much.

![J DZ$@}YQ8@) EIEDMGOJ@U](https://github.com/vuejs/vitepress/assets/10327999/b6e110c0-e5c6-4e95-bb3d-6bb4694aefce)

```
{
  "type": "module",
  "license": "BSD-3-Clause",
  "dependencies": {
    "vitepress": "1.2.2",
	"vite-plugin-vitepress-auto-nav": "2.3.2",
	"markdown-it-footnote": "4.0.0",
	"markdown-it-mathjax3": "git+https://git@github.com:taophilosophy/markdown-it-mathjax3.git",
	"markdown-it-task-checkbox": "1.0.6"
  },
  "scripts": {
    "docs:dev": "vitepress dev docs",
    "docs:build": "vitepress build docs",
    "docs:preview": "vitepress preview docs"
  }
}
```

```
import {
	defineConfig
} from 'vitepress';
import autoNav from "vite-plugin-vitepress-auto-nav";
import footnote from 'markdown-it-footnote';
import mathjax3 from 'markdown-it-mathjax3';
import taskLists from 'markdown-it-task-checkbox';

const customElements = [
  'mjx-container',
  'mjx-assistive-mml',
  'math',
  'maction',
  'maligngroup',
  'malignmark',
  'menclose',
  'merror',
  'mfenced',
  'mfrac',
  'mi',
  'mlongdiv',
  'mmultiscripts',
  'mn',
  'mo',
  'mover',
  'mpadded',
  'mphantom',
  'mroot',
  'mrow',
  'ms',
  'mscarries',
  'mscarry',
  'mscarries',
  'msgroup',
  'mstack',
  'mlongdiv',
  'msline',
  'mstack',
  'mspace',
  'msqrt',
  'msrow',
  'mstack',
  'mstack',
  'mstyle',
  'msub',
  'msup',
  'msubsup',
  'mtable',
  'mtd',
  'mtext',
  'mtr',
  'munder',
  'munderover',
  'semantics',
  'math',
  'mi',
  'mn',
  'mo',
  'ms',
  'mspace',
  'mtext',
  'menclose',
  'merror',
  'mfenced',
  'mfrac',
  'mpadded',
  'mphantom',
  'mroot',
  'mrow',
  'msqrt',
  'mstyle',
  'mmultiscripts',
  'mover',
  'mprescripts',
  'msub',
  'msubsup',
  'msup',
  'munder',
  'munderover',
  'none',
  'maligngroup',
  'malignmark',
  'mtable',
  'mtd',
  'mtr',
  'mlongdiv',
  'mscarries',
  'mscarry',
  'msgroup',
  'msline',
  'msrow',
  'mstack',
  'maction',
  'semantics',
  'annotation',
  'annotation-xml',
  'mjx-c',
  'mjx-mstyle',
  'mjx-mspace',
  'mjx-mover',
  'mjx-base',
  'mjx-over',
  'mjx-texatom',
  'mjx-mfrac',
  'mjx-frac',
  'mjx-dbox',
  'mjx-dtable',
  'mjx-row',
  'mjx-den',
  'mjx-dstrut',
  'mjx-line',
  'mjx-num',
  'mjx-mrow',
  'mjx-msqrt',
  'mjx-sqrt',
  'mjx-box',
  'mjx-surd',
  'mjx-nstrut',
  'mjx-msup',
  'mjx-script',
  'mjx-math',
  'mjx-mn',
  'mjx-mo',
  'mjx-mi',
];

export default defineConfig({
	sitemap: {
		hostname: 'https://taophilosophy.github.io',
	},
	lang: 'zh-CN',
	lastUpdated: true,
	title: "斯坦福哲学百科全书简体中文版",
	description: "斯坦福哲学百科全书",
	metaChunk: true,
	markdown: {
		image: {
			lazyLoading: true
		},
		config(md) {
			md.use(footnote);
			md.use(mathjax3, {
               tex: {tags: 'ams'},
               loader: {load: ["input/tex", "output/chtml"]},
			});
			md.use(taskLists, {
				disabled: true,
				divWrap: false,
				divClass: 'checkbox',
				idPrefix: 'cbx_',
				ulClass: 'task-list',
				liClass: 'task-list-item',
			});
			md.renderer.rules.footnote_anchor = function render_footnote_anchor(tokens, idx, options, env, slf) {
				let id = slf.rules.footnote_anchor_name?.(tokens, idx, options, env, slf)
				if (tokens[idx].meta.subId > 0) {
					id += ':' + tokens[idx].meta.subId
				}
				return ' <a href="#fnref' + id + '" class="footnote-backref">🔼</a>'
			};
			const defaultCodeInline = md.renderer.rules.code_inline;

			md.renderer.rules.code_inline = (tokens, idx, options, env, self) => {
				tokens[idx].attrSet('v-pre', '');
				return defaultCodeInline(tokens, idx, options, env, self);
			};
		}

	},
   vue: {
    template: {
      compilerOptions: {
        isCustomElement: (tag) => customElements.includes(tag),
      },
    },
  },
	head: [
		['link', {
			rel: 'icon',
			href: '/favicon.ico'
		}],
		['meta', {
			name: 'keywords',
			content: ''
		}],
		[
			'script',
			{
				async: '',
				src: 'https://www.googletagmanager.com/gtag/js?id=TAG_ID'
			}
		],
		[
			'script',
			{},
			`window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());
      gtag('config', 'TAG_ID');`
		],
	],
	rewrites: {
		'README.md': 'index.md',
	},
	themeConfig: {
		siteTitle: '斯坦福哲学百科全书',
		langMenuLabel: '多语言',
		returnToTopLabel: '回到顶部',
		darkModeSwitchLabel: '主题',
		lightModeSwitchTitle: '切换到浅色模式',
		darkModeSwitchTitle: '切换到深色模式',
		docFooter: {
			prev: '上一页',
			next: '下一页'
		},

		logo: {
			src: '/logo.svg',
			width: 24,
			height: 24
		},
		nav: [{
				text: '斯坦福哲学百科全书简体中文版-分类目录',
				link: 'mu-lu-s.md'
			},
			{
				text: '原版目录（A-Z）',
				link: 'SUMMARY.md'
			},
		],
		base: '/',
		editLink: {
			text: '在 GitHub 上编辑此页面',
			pattern: 'https://github.com/taophilosophy/SEP-CN/edit/main/:path'
		},
		outline: {
			label: '此页目录',
			level: 'deep'
		},
		sidebarMenuLabel: '目录',
		lastUpdated: {
			text: '最后更新于',
			formatOptions: {
				dateStyle: 'short',
				timeStyle: 'medium'
			}
		},
	},


	vite: {
		plugins: [
			autoNav({
				summary: {
					target: "docs/mu-lu-s.md",
					collapsed: false,
				}
			})
		],
	},
})
```
`custom.css`
```
/* 脚注加粗，非斜体 */
.footnote-item {
  font-style: normal;
  font-weight: bold;
}
/* 去掉链接下划线 */
.vp-doc a {
  text-decoration: none;
}
@media (min-width: 1440px) {
/* 导航栏 */
  .VPNavBar .content {
    /* 导航栏左侧最远至，为 0 时到达最左侧，覆盖原来的位置 */
    padding-left: 10vw !important;
    /* 导航栏右侧最远至，为 0 时到达最右侧，覆盖原来的位置 */
    padding-right: 0vw !important;
  }
  /* 导航栏标题左、右侧缩进 */
.VPNavBar {
  padding-left:  5% !important;
  padding-right: 5% !important;
}
  /* 右 TOC */
.aside {
/* 右 TOC 向左缩进（距离右边框）*/
  max-width: 5vw !important;
/* 右 TOC 向右缩进（与文本的距离）*/
  padding-left: 2vw !important;
}
/* 导航栏标题 */
  .VPNavBar .title {
    /* 导航栏标题左侧缩进 */
    padding-left: 1.5vw !important;
  }
  .aside-container{
  /* 右 TOC 文本宽度*/
    width: 16vw !important;
  }
  .outline-link{
  /* 右 TOC 文本换行*/
   white-space: normal !important;
  }
/* 左侧边栏 */
    .VPSidebar {
    /* 左侧边栏左缩进 */
    padding-left: 2vw !important;
    /* 左侧边栏右缩进 */
    width: 15vw !important;
  }
  /* 文本左、右缩进（整体） */
  .VPContent.has-sidebar {
    padding-left: 15% !important;
    padding-right: 15% !important;
  }
 :root{
   /* 左侧边栏文字宽度 */
    --vp-sidebar-width: 14.5vw !important;
  }
}
@media (min-width: 960px) {
  .VPDoc.has-aside .content-container {
  max-width: 100% !important;
}
}
mjx-container {
  display: inline-block;
  margin: auto 2px -2px;
}

mjx-container > svg {
  margin: auto;
  display: inline-block;
}
````

`index.js`
```
import DefaultTheme from 'vitepress/theme'
import './custom.css'

export default DefaultTheme
```

See also <https://taophilosophy.github.io/>

**Comment by brc-dd** [comment]
Ah, it's not that simple. That fork won't work properly. You need to insert generated CSS. Those fonts are these. But other CSS is also missing.

```css
@font-face {
  font-family: 'MJXZERO';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Zero.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Main-Regular.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-B';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Main-Bold.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-I';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Math-Italic.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-MI';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Main-Italic.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-BI';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Math-BoldItalic.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-S1';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Size1-Regular.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-S2';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Size2-Regular.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-S3';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Size3-Regular.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-S4';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Size4-Regular.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-A';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_AMS-Regular.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-C';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Calligraphic-Regular.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-CB';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Calligraphic-Bold.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-FR';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Fraktur-Regular.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-FRB';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Fraktur-Bold.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-SS';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_SansSerif-Regular.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-SSB';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_SansSerif-Bold.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-SSI';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_SansSerif-Italic.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-SC';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Script-Regular.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-T';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Typewriter-Regular.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-V';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Vector-Regular.woff')
    format('woff');
}

@font-face {
  font-family: 'MJXTEX-VB';
  src: url('https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2/MathJax_Vector-Bold.woff')
    format('woff');
}
```

**Reply by ykla** [reply]
Thanks, for my test, it seems just need:

```
	head: [
		[
			'script',
			{
				async: '',
				src: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/3.2.2/es5/tex-mml-chtml.js',
			}
		],
	],
```

However, I found that this could not be copied either, and consulting the documentation I found that I may need to use `mml]/mml3` output. <https://docs.mathjax.org/en/latest/input/mathml.html>

**Reply by brc-dd** [reply]
chtml is supported in this one for vuepress - https://theme-hope.vuejs.press/ (https://mdit-plugins.github.io/mathjax.html)

someone might be able to port that plugin to vitepress 👀

**Reply by ykla** [reply]
After I merged the code I mentioned above, regarding the chtml output I have successfully implemented it. I am working on outputting in mml3. <https://github.com/taophilosophy/markdown-it-mathjax3/commit/26fd2dd37036c148d634eb89a646108a77e0b4ce> works with chtml output. But what I really need is the output in mml3. But implementing chtml output is a prerequisite for mml3 output. Currently chtml output has been implemented. But I don't know how to implement mml3 output, there is so little code to refer to, and I can hardly find any open source implementation of mathjax mml3. The advantage of exporting in mml3 is that the code can be copied. The other two types of output are not reproducible on a web page.
